xcm/xcm-builder/src/barriers.rs

```diff
@@ -210,6 +210,85 @@ impl<
 	}
 }
 
+/// Allows the message to be prepended with a single `SetTopic` instruction, requiring some inner
+/// barrier to pass on the rest of the message.
+pub struct AllowSetTopic<InnerBarrier>(PhantomData<InnerBarrier>);
+impl<InnerBarrier: ShouldExecute> ShouldExecute for AllowSetTopic<InnerBarrier> {
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		instructions: &mut [Instruction<Call>],
+		max_weight: Weight,
+		weight_credit: &mut Weight,
+	) -> Result<(), ProcessMessageError> {
+		log::trace!(
+			target: "xcm::barriers",
+			"AllowSetTopic origin: {:?}, instructions: {:?}, max_weight: {:?}, weight_credit: {:?}",
+			origin, instructions, max_weight, weight_credit,
+		);
+		let skipped = if let Some(SetTopic(..)) = instructions.first() { 1 } else { 0 };
+		InnerBarrier::should_execute(
+			&origin,
+			&mut instructions[skipped..],
+			max_weight,
+			weight_credit,
+		)
+	}
+}
+
+/// Requires the message to be prepended with a single `SetTopic` instruction, as well as some inner
+/// barrier to pass on the rest of the message.
+pub struct RequireSetTopic<InnerBarrier>(PhantomData<InnerBarrier>);
+impl<InnerBarrier: ShouldExecute> ShouldExecute for RequireSetTopic<InnerBarrier> {
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		instructions: &mut [Instruction<Call>],
+		max_weight: Weight,
+		weight_credit: &mut Weight,
+	) -> Result<(), ProcessMessageError> {
+		log::trace!(
+			target: "xcm::barriers",
+			"AllowSetTopic origin: {:?}, instructions: {:?}, max_weight: {:?}, weight_credit: {:?}",
+			origin, instructions, max_weight, weight_credit,
+		);
+		if let Some(SetTopic(..)) = instructions.first() {
+			InnerBarrier::should_execute(&origin, &mut instructions[1..], max_weight, weight_credit)
+		} else {
+			Err(ProcessMessageError::Unsupported)
+		}
+	}
+}
+
+/// Wrapper router which prepends a `SetTopic` instruction to the message prior to sending with a
+/// universally unique topic, and returns this value from a successful `deliver`.
+///
+/// This is designed to be at the top-level of any routers, since it will always mutate the
+/// passed `message` reference into a `None`. Don't try to combine it within a tuple except as the
+/// last element.
+pub struct WithUniqueTopic<Inner>(PhantomData<Inner>);
+impl<Inner: xcm::latest::SendXcm> xcm::latest::SendXcm for WithUniqueTopic<Inner> {
```

Comment:
Looks kind of weird to have a router implementation appear in barrier.rs. Aren't there better places for this?